### PR TITLE
TICK-375 add prox binarsity

### PR DESCRIPTION
**Pull request by SimonBussy**

### CODE REVIEW

File: tick/inference/base/learner_glm.py [hunk 1]
```diff
@@ -53,21 +53,29 @@ class LearnerGLM(LearnerOptim):
         Record history information when ``n_iter`` (iteration number) is
         a multiple of ``record_every``
 
+    random_state : int seed, RandomState instance, or None (default)
```
Comment (Mbompr) [inline tick/inference/base/learner_glm.py]: did you add this in this task?

File: tick/inference/base/learner_glm.py [hunk 2]
```diff
@@ -53,21 +53,29 @@ class LearnerGLM(LearnerOptim):
...
+    Other Parameters
+    ----------------
     sdca_ridge_strength : `float`, default=1e-3
         It controls the strength of the additional ridge penalization. Used in
         'sdca' solver
 
     elastic_net_ratio : `float`, default=0.95
         Ratio of elastic net mixing parameter with 0 <= ratio <= 1.
-        For ratio = 0 this is ridge (L2) regularization
+        For ratio = 0 this is ridge (L2 squared) regularization
```
Comment (Mbompr) [inline tick/inference/base/learner_glm.py]: same here

File: tick/inference/tests/coxreg_test.py [hunk 3]
```diff
@@ -52,10 +53,29 @@ def test_CoxRegression_fit(self):
...
 
         for penalty in self.penalties:
+
+            if penalty == 'binarsity':
+                # binarize features
+                n_cuts = 3
+                binarizer = FeaturesBinarizer(n_cuts=n_cuts)
+                X = binarizer.fit_transform(features)
+            else:
+                X = features
```
Comment (Mbompr) [inline tick/inference/tests/coxreg_test.py]: it's better to keep more explicit `feature` name

File: tick/inference/base/learner_optim.py [hunk 4]
```diff
@@ -153,6 +165,8 @@ def __init__(self, penalty='l2', C=1e3, solver="svrg", step=None,
         if extra_prox_kwarg is None:
             extra_prox_kwarg = {}
         self._prox_obj = self._construct_prox_obj(penalty, elastic_net_ratio,
```
Comment (Mbompr) [inline tick/inference/base/learner_optim.py]: You should store `block_start` and `block_length` somewhere in the learner, such that people can retrieve their values by calling `learner.block_start`.
I suggest you do something like what we have done with `elastic_net_ratio` including the `@elastic_net_ratio.setter` which will raise a warning if one of these parameter is given when penalty is not binarsity

File: tick/inference/tests/logreg_test.py [hunk 5]
```diff
@@ -38,18 +40,32 @@ def test_LogisticRegression_fit(self):
...
         for fit_intercept in [True, False]:
             for penalty in penalties:
+
+                if penalty == 'binarsity':
+                    # binarize features
+                    n_cuts = 3
+                    binarizer = FeaturesBinarizer(n_cuts=n_cuts)
+                    X = binarizer.fit_transform(features)
+                else:
+                    X = features
```
Comment (Mbompr) [inline tick/inference/tests/logreg_test.py]: X should be called features

File: tick/optim/prox/prox_binarsity.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+import numpy as np
+from .base import Prox
+from .build.prox import ProxBinarsity as _ProxBinarsity
+
+
+class ProxBinarsity(Prox):
+    """Proximal operator of binarsity, namely total-variation
+    regularization followed by centering within sub-blocks of the
+    coefficients. Blocks (non-overlapping) are specified by the
+    ``blocks_start`` and ``blocks_length`` parameters.
```
Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: Do you check anywhere in the code that the blocks are indeed not overlapping each other?

File: tick/optim/prox/prox_binarsity.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
```
Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: Maybe should you detail a bit more how this prox is different from TV and what a block is.

File: tick/optim/prox/prox_binarsity.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    Attributes
+    ----------
+    n_blocks : `int`
+        Number of blocks
+
+    Notes
+    -----
+    Uses the fast-TV algorithm described in:
+
+    * "A Direct Algorithm for 1D Total Variation Denoising"
```
Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: Original paper should be cited with a link (example in tick/inference/hawkes_em.py)

File: tick/optim/prox/prox_binarsity.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        if blocks_length.dtype is not np.uint64:
+            blocks_length = blocks_length.astype(np.uint64)
+
+        if blocks_start.shape != blocks_length.shape:
+            raise ValueError("``blocks_start`` and ``blocks_length`` "
+                             "must have the same size")
+
+        self.strength = strength
+        self.positive = positive
+        self.n_blocks = blocks_start.shape[0]
```
Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: this should be a property

```
@property
def n_blocks(self):
    return self.blocks_start.shape[0]
```

This way it will always be synchronized with blocks_start

Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: Or you can get this value from C++ (even better)

```
@property
def n_blocks(self):
    return self._prox.get_n_blocks()
```

File: tick/optim/prox/src/prox_binarsity.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+#include "prox_tv.h"
+#include "prox_binarsity.h"
+
+
+ProxBinarsity::ProxBinarsity(double strength,
+                             SArrayULongPtr blocks_start,
+                             SArrayULongPtr blocks_length,
+                             bool positive)
+        : Prox(strength)
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: write ` : Prox(strength), ready(false)`

File: tick/optim/prox/src/prox_binarsity.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+{
+    this->blocks_start = blocks_start;
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: Check that block_start and block_length have the same size, even if this has been checked in Python (always check in C++ as it could lead to a segfault)

File: tick/optim/prox/src/prox_binarsity.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+    SArrayULongPtr blocks_start;
+    SArrayULongPtr blocks_length;
+
+    // A vector that contains the prox for each block
+    std::vector<ProxTV> blocks_prox;
+
+    // Tells us if the prox is ready (with correctly allocated sub-prox for each blocks).
+    // This is mainly necessary when the user changes the range from python
+    bool ready;
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.h]: I would rename it to something more meaningful such as 
`prox_tv_synchronized`

File: tick/optim/prox/src/prox_binarsity.h [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    void prepare();
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.h]: I would rename it to something more meaningful such as 
`synchronize_prox_tv`

File: tick/inference/base/learner_optim.py (hunk 4 above)
Comment (SimBussy) [inline tick/inference/base/learner_optim.py]: done!

File: tick/inference/tests/coxreg_test.py (hunk 3 above)
Comment (SimBussy) [inline tick/inference/tests/coxreg_test.py]: ok

File: tick/inference/tests/logreg_test.py (hunk 5 above)
Comment (SimBussy) [inline tick/inference/tests/logreg_test.py]: ok

File: tick/optim/prox/prox_binarsity.py (hunk 6 above)
Comment (SimBussy) [inline tick/optim/prox/prox_binarsity.py]: no overlapping test added

File: tick/optim/prox/prox_binarsity.py (hunk 8 above)
Comment (SimBussy) [inline tick/optim/prox/prox_binarsity.py]: ok

File: tick/optim/prox/prox_binarsity.py (hunk 9 above)
Comment (SimBussy) [inline tick/optim/prox/prox_binarsity.py]: done

File: tick/optim/prox/src/prox_binarsity.cpp (hunk 11 above)
Comment (SimBussy) [inline tick/optim/prox/src/prox_binarsity.cpp]: ok

File: tick/optim/prox/src/prox_binarsity.h (hunk 12 above)
Comment (SimBussy) [inline tick/optim/prox/src/prox_binarsity.h]: ok

File: tick/optim/prox/src/prox_binarsity.h (hunk 13 above)
Comment (SimBussy) [inline tick/optim/prox/src/prox_binarsity.h]: ok

File: tick/inference/base/learner_glm.py (hunk 1 above)
Comment (Mbompr) [inline tick/inference/base/learner_glm.py]: It should be moved to Other Parameters after `elastic_net_ratio` (it should be in the same order as in the constructor)

File: tick/inference/base/learner_optim.py [hunk 14]
```diff
@@ -52,21 +53,30 @@ class LearnerOptim(ABC, Base):
         Record history information when ``n_iter`` (iteration number) is
         a multiple of ``record_every``
 
+    random_state : int seed, RandomState instance, or None (default)
+        The seed that will be used by stochastic solvers. Used in 'sgd',
+        'svrg', and 'sdca' solvers
```
Comment (Mbompr) [inline tick/inference/base/learner_optim.py]: It should be moved to Other Parameters after `elastic_net_ratio` (it should be in the same order as in the constructor)

File: tick/inference/base/learner_glm.py [hunk 15]
```diff
@@ -53,21 +53,29 @@ class LearnerGLM(LearnerOptim):
...
-
-    random_state : int seed, RandomState instance, or None (default)
-        The seed that will be used by stochastic solvers. Used in 'sgd',
-        'svrg', and 'sdca' solvers
+        
+    blocks_start : `numpy.array`, shape=(n_features,), default=None
+        The indices of the first column of each binarized feature blocks
+        
+    blocks_length : `numpy.array`, shape=(n_features,), default=None
+        The length of each binarized feature blocks
```
Comment (Mbompr) [inline tick/inference/base/learner_glm.py]: Someone who is using learner does not know much about prox binarsity. You should be very explicit about these two parameters and specify that they will only be used for prox binarsity

File: tick/inference/base/learner_optim.py [hunk 16]
```diff
@@ -52,21 +53,30 @@ class LearnerOptim(ABC, Base):
...
-
-    random_state : int seed, RandomState instance, or None (default)
-        The seed that will be used by stochastic solvers. Used in 'sgd',
-        'svrg', and 'sdca' solvers
+        
+    blocks_start : `numpy.array`, shape=(n_features,), default=None
+        The indices of the first column of each binarized feature blocks
+        
+    blocks_length : `numpy.array`, shape=(n_features,), default=None
+        The length of each binarized feature blocks
```
Comment (Mbompr) [inline tick/inference/base/learner_optim.py]: Someone who is using learner does not know much about prox binarsity. You should be very explicit about these two parameters and specify that they will only be used for prox binarsity

File: tick/inference/cox_regression.py [hunk 17]
```diff
@@ -206,7 +221,8 @@ def score(self, features=None, times=None, censoring=None):
                 else:
                     features, times, censoring = self._all_safe(features, times,
                                                                 censoring)
-                    model = ModelCoxRegPartialLik().fit(features, times, censoring)
+                    model = ModelCoxRegPartialLik().fit(features, times,
+                                                        censoring)
```
Comment (Mbompr) [inline tick/inference/cox_regression.py]: no need to break line

File: tick/inference/tests/coxreg_test.py [hunk 18]
```diff
@@ -52,10 +53,29 @@ def test_CoxRegression_fit(self):
...
+                 -0.75580671, 0.0593833, -0.01433773, 0.01077109, -0.05581666])
         }
 
         for penalty in self.penalties:
+
+            if penalty == 'binarsity':
+                # binarize features
+                n_cuts = 3
+                binarizer = FeaturesBinarizer(n_cuts=n_cuts)
+                features = binarizer.fit_transform(features_)
```
Comment (Mbompr) [inline tick/inference/tests/coxreg_test.py]: this `features_` is not super explicit. I would rather call it `raw_features` or simply call it `features` and override by `features = binarizer.fit_transform(features)`

File: tick/inference/tests/coxreg_test.py [hunk 19]
```diff
@@ -65,9 +85,17 @@ def test_CoxRegression_fit(self):
                 if penalty != 'none':
                     solver_kwargs['C'] = 50
 
+                if penalty == 'binarsity':
+                    solver_kwargs['blocks_start'] = binarizer.feature_indices[
+                                                    :-1, ]
```
Comment (Mbompr) [inline tick/inference/tests/coxreg_test.py]: does not fit in one line?

File: tick/inference/tests/coxreg_test.py [hunk 20]
```diff
@@ -65,9 +85,17 @@ def test_CoxRegression_fit(self):
...
+                    solver_kwargs['blocks_length'] = binarizer.n_values
+
                 learner = CoxRegression(**solver_kwargs)
                 learner.fit(features, times, censoring)
 
+                if penalty == 'binarsity':
+                    print(learner.coeffs)
```
Comment (Mbompr) [inline tick/inference/tests/coxreg_test.py]: you've forgot this

File: tick/inference/base/learner_optim.py [hunk 21]
```diff
@@ -172,6 +185,12 @@ def __init__(self, penalty='l2', C=1e3, solver="svrg", step=None,
                         penalty == 'elasticnet':
             self.elastic_net_ratio = elastic_net_ratio
 
+        if 'blocks_start' in self._actual_kwargs and penalty == 'binarsity':
```
Comment (Mbompr) [inline tick/inference/base/learner_optim.py]: should be a `or` not a `and`.
The idea is that if user set parameter `blocks_start` with a l2 penalty it should get a warning

File: tick/inference/tests/linear_regression_test.py [hunk 22]
```diff
@@ -55,7 +55,9 @@ def test_LinearRegression_fit(self):
                                        fit_intercept=fit_intercept,
                                        solver=solver,
                                        penalty=penalty,
-                                       tol=1e-10)
+                                       tol=1e-10,
+                                       blocks_start=[0],
+                                       blocks_length=[1])
```
Comment (Mbompr) [inline tick/inference/tests/linear_regression_test.py]: You cannot add `blocks_start=[0], blocks_length=[1]` just because it's convenient :)

File: tick/optim/prox/src/prox_binarsity.cpp [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+                             SArrayULongPtr blocks_start,
+                             SArrayULongPtr blocks_length,
+                             ulong start,
+                             ulong end, bool positive)
+        : Prox(strength, start, end, positive) {
+    // blocks_start and blocks_end have the same size
+    n_blocks = blocks_start->size();
+    if (n_blocks != blocks_length->size()) {
+        throw std::invalid_argument("blocks_start and blocks_length must have the same size");
+    }
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: this check should be done in `set_blocks_start` and in `set_blocks_length` and you should call these functions in the constructor

Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: Otherwise someone can start with blocks of same length and modify one afterwars which would lead to a segfault

File: tick/optim/prox/src/prox_binarsity.h [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        this->end = end;
+    }
+
+    inline virtual void set_blocks_start(SArrayULongPtr blocks_start) {
+        this->blocks_start = blocks_start;
+    }
+
+    inline virtual void set_blocks_length(SArrayULongPtr blocks_length) {
+        this->blocks_length = blocks_length;
+    }
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.h]: shouldn't we set `prox_tv_synchronized` to false in each setter ?

File: tick/optim/prox/prox_binarsity.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+from .base import Prox
+from .build.prox import ProxBinarsity as _ProxBinarsity
+
+
+class ProxBinarsity(Prox):
+    """Proximal operator of binarsity, namely total-variation
+    regularization applied on a coefficient vector being a concatenation of
+    multiple coefficient vectors corresponding to blocks, followed by centering
+    within sub-blocks. Blocks (non-overlapping) are specified by the
+    ``blocks_start`` and ``blocks_length`` parameters.
```
Comment (Mbompr) [inline tick/optim/prox/prox_binarsity.py]: Can you add that it simply is a succession of prox tv on different intervals?

File: tick/optim/prox/src/prox_binarsity.cpp [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+    this->blocks_start = blocks_start;
+    this->blocks_length = blocks_length;
+    this->positive = positive;
+    // The object is not ready (synchronize_prox_tv has not been called)
+    prox_tv_synchronized = false;
+}
+
+
+void ProxBinarsity::synchronize_prox_tv() {
+    if (!prox_tv_synchronized) {
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: I would do the opposite:
if `synchronize_prox_tv` is called, you synchronize it without looking at`prox_tv_synchronized`. 
If `value` or `call` is called it looks at `prox_tv_synchronized` to know if it should call `synchronize_prox_tv`

File: tick/optim/prox/tests/prox_binarsity_test.py [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+        val = 0
+        for j, d_j in enumerate(blocks_length):
+            start = blocks_start[j]
+            val += np.abs(coeffs[start + 1:start + d_j]
+                          - coeffs[start:start + d_j - 1]).sum()
+        val *= l_binarsity
+
+        self.assertAlmostEqual(prox.value(coeffs), val, delta=1e-15)
+        assert_almost_equal(prox.call(coeffs, step=t), out, decimal=7)
+
```
Comment (Mbompr) [inline tick/optim/prox/tests/prox_binarsity_test.py]: Can you add few tests to check you exceptions (overlapping blocks, negative length, different size in constructor or after changing one block parameter etc.) :)

File: tick/inference/base/learner_glm.py (hunk 1 above)
Comment (SimonBussy) [inline tick/inference/base/learner_glm.py]: ok

File: tick/inference/base/learner_optim.py (hunk 14 above)
Comment (SimonBussy) [inline tick/inference/base/learner_optim.py]: done

File: tick/inference/base/learner_glm.py (hunk 15 above)
Comment (SimonBussy) [inline tick/inference/base/learner_glm.py]: ok

File: tick/inference/base/learner_optim.py (hunk 16 above)
Comment (SimonBussy) [inline tick/inference/base/learner_optim.py]: ok

File: tick/inference/tests/coxreg_test.py (hunk 18 above)
Comment (SimonBussy) [inline tick/inference/tests/coxreg_test.py]: ok for `raw_features`. The problem of overriding `features` is that if the binarsity penalty would have to be the last tested one in the for loop

File: tick/inference/tests/coxreg_test.py (hunk 19 above)
Comment (SimonBussy) [inline tick/inference/tests/coxreg_test.py]: no

File: tick/inference/cox_regression.py (hunk 17 above)
Comment (SimonBussy) [inline tick/inference/cox_regression.py]: does not fit in one line

File: tick/inference/tests/coxreg_test.py (hunk 20 above)
Comment (SimonBussy) [inline tick/inference/tests/coxreg_test.py]: indeed!

File: tick/inference/tests/linear_regression_test.py (hunk 22 above)
Comment (SimonBussy) [inline tick/inference/tests/linear_regression_test.py]: ok

File: tick/inference/tests/coxreg_test.py (hunk 18 above)
Comment (Mbompr) [inline tick/inference/tests/coxreg_test.py]: Oh, I didn't realize that indeed :)

File: tick/optim/prox/src/prox_binarsity.cpp (hunk 23 above)
Comment (SimonBussy) [inline tick/optim/prox/src/prox_binarsity.cpp]: done

File: tick/optim/prox/src/prox_binarsity.h (hunk 24 above)
Comment (SimonBussy) [inline tick/optim/prox/src/prox_binarsity.h]: definitely!

File: tick/optim/prox/prox_binarsity.py (hunk 25 above)
Comment (SimonBussy) [inline tick/optim/prox/prox_binarsity.py]: ok

File: tick/optim/prox/src/prox_binarsity.cpp (hunk 26 above)
Comment (SimonBussy) [inline tick/optim/prox/src/prox_binarsity.cpp]: makes more sense!

File: tick/optim/prox/tests/prox_binarsity_test.py (hunk 27 above)
Comment (SimonBussy) [inline tick/optim/prox/tests/prox_binarsity_test.py]: done

File: tick/inference/cox_regression.py [hunk 28]
```diff
@@ -52,6 +52,12 @@ class CoxRegression(LearnerOptim):
         Record history information when ``n_iter`` (iteration number) is
         a multiple of ``record_every``
 
+    Other Parameters
+    ----------------
+    sdca_ridge_strength : `float`, default=1e-3
```
Comment (Mbompr) [inline tick/inference/cox_regression.py]: this model does not use sdca...

File: tick/inference/linear_regression.py [hunk 29]
```diff
@@ -65,6 +67,12 @@ class LinearRegression(LearnerGLM):
         The seed that will be used by stochastic solvers. Used in 'sgd',
         'svrg', and 'sdca' solvers
 
+    blocks_start : `numpy.array`, shape=(n_features,), default=None
+        The indices of the first column of each binarized feature blocks
+
+    blocks_length : `numpy.array`, shape=(n_features,), default=None
+        The length of each binarized feature blocks
+
```
Comment (Mbompr) [inline tick/inference/linear_regression.py]: Please provide long descriptions, this applies to all learners
```
+    blocks_start : `numpy.array`, shape=(n_features,), default=None
--
  |   |   | +        The indices of the first column of each binarized feature blocks. It
  |   |   | +        corresponds to the ``feature_indices`` property of the
  |   |   | +        ``FeaturesBinarizer`` preprocessing.
  |   |   | +        Used in 'binarsity' penalty
  |   |   | +
  |   |   | +    blocks_length : `numpy.array`, shape=(n_features,), default=None
  |   |   | +        The length of each binarized feature blocks. It corresponds to the
  |   |   | +        ``n_values`` property of the ``FeaturesBinarizer`` preprocessing.
  |   |   | +        Used in 'binarsity' penalty


```
